src/backend/metal/src/device.rs

```diff
@@ -800,28 +800,64 @@ impl hal::Device<Backend> for Device {
     fn destroy_sampler(&self, _sampler: n::Sampler) {
     }
 
-    fn map_memory(&self, _: &n::Memory, _: Range<u64>) -> Result<*mut u8, mapping::Error> {
-        unimplemented!()
+    fn map_memory(&self, memory: &n::Memory, range: Range<u64>) -> Result<*mut u8, mapping::Error> {
+        let allocations = memory.allocations.lock().unwrap();
+        let mut mapping = memory.mapping.lock().unwrap();
+
+        assert!(mapping.is_none(), "Only one mapping per `Memory` at a time is allowed");
+        let buffers = allocations.find(range.clone());
+
+        assert_eq!(buffers.len(), 1, "Only mapping range within single buffer is alowed for now");
+        let (buffer_range, buffer) = buffers.into_iter().next().unwrap();
+
+        debug_assert!(range.start >= buffer_range.start);
+        debug_assert!(range.end <= buffer_range.end);
+        debug_assert_eq!(buffer.length(), buffer_range.end - buffer_range.start);
+
+        let offset = range.start - buffer_range.start;
+        let length = range.end - range.start;
+        let ptr = unsafe {
+            (buffer.contents() as *mut u8).offset(offset as isize)
+        };
+
+        *mapping = Some(n::MemoryMapping {
+            range,
+            buffer,
+            location: offset as _,
+            length: length as _,
+        });
+        Ok(ptr)
     }
 
-    fn unmap_memory(&self, _: &n::Memory) {
-        unimplemented!()
+    fn unmap_memory(&self, memory: &n::Memory) {
+        memory.mapping.lock().unwrap().take();
     }
 
-    fn flush_mapped_memory_ranges<'a, I>(&self, _: I)
+    fn flush_mapped_memory_ranges<'a, I>(&self, iter: I)
     where
         I: IntoIterator,
         I::Item: Borrow<(&'a n::Memory, Range<u64>)>,
     {
-        unimplemented!()
+        for item in iter.into_iter() {
+            let (memory, ref range) = *item.borrow();
+            let mut mapping = memory.mapping.lock().unwrap();
+            assert!(mapping.is_some());
+            let mapping = mapping.as_ref().unwrap();
```

Comment:
the assert above is covered by this `unwrap`

Reply:
Correct.